source/extensions/retry/host/other_hosts/other_hosts.h

```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#include "envoy/upstream/retry.h"
+#include "envoy/upstream/upstream.h"
+
+namespace Envoy {
+class OtherHostsRetryPredicate : public Upstream::RetryHostPredicate {
+public:
+  bool shouldSelectAnotherHost(const Upstream::Host& candidate_host) override {
+    return attempted_hosts_.find(candidate_host.address()->asString()) != attempted_hosts_.end();
+  }
+  void onHostAttempted(Upstream::HostDescriptionConstSharedPtr attempted_host) override {
+    attempted_hosts_.insert(attempted_host->address()->asString());
+  }
+
+private:
+  std::unordered_set<std::string> attempted_hosts_;
```

Comment:
@snowp I'm not sure of my comment came through here or not: I suspect that a vector or fixed size array initialized to the number of max retries would end up being faster here and use less memory given the generally tiny number of allowed retries. Something to think about / TODO in the future.